### ISSUE TITLE
sdist needs to be updated to include content databases

### ISSUE THREAD

**Issue by MCGallaspy** [learningequality/ka-lite#4653]
While tackling an issue with the Windows installer build process, I encountered the following issue. Currently on `develop` when I run `make dist` and install the output with `pip install <outfile>`, when I try to start the server I get the following error:

```sh
INFO:kalite:Annotating content models for language: en, channel: khan in databas
e located at C:\Users\Mike\.kalite\content\content_khan_en.sqlite
Traceback (most recent call last):
  File "C:\Python27\share\kalite\python-packages\django\core\management\base.py"
, line 224, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "C:\Python27\share\kalite\python-packages\django\core\management\base.py"
, line 263, in execute
    output = self.handle(*args, **options)
  File "C:\Python27\lib\site-packages\kalite\distributed\management\commands\ini
tialize_kalite.py", line 48, in handle
    self.setup_server_if_needed()
  File "C:\Python27\lib\site-packages\kalite\distributed\management\commands\ini
tialize_kalite.py", line 34, in setup_server_if_needed
    call_command("setup", interactive=False)
  File "C:\Python27\share\kalite\python-packages\django\core\management\__init__
.py", line 161, in call_command
    return klass.execute(*args, **defaults)
  File "C:\Python27\share\kalite\python-packages\django\core\management\base.py"
, line 263, in execute
    output = self.handle(*args, **options)
  File "C:\Python27\lib\site-packages\kalite\distributed\management\commands\set
up.py", line 491, in handle
    call_command("annotate_content_items")
  File "C:\Python27\share\kalite\python-packages\django\core\management\__init__
.py", line 161, in call_command
    return klass.execute(*args, **defaults)
  File "C:\Python27\share\kalite\python-packages\django\core\management\base.py"
, line 263, in execute
    output = self.handle(*args, **options)
  File "C:\Python27\lib\site-packages\kalite\contentload\management\commands\ann
otate_content_items.py", line 51, in handle
    annotate_content_models(database_path=database_path, channel=channel, langua
ge=language)
  File "C:\Python27\lib\site-packages\kalite\topic_tools\content_models.py", lin
e 103, in wrapper
    output = function(*args, **kwargs)
  File "C:\Python27\lib\site-packages\kalite\topic_tools\content_models.py", lin
e 398, in annotate_content_models
    content_models = get_content_items(ids=ids, channel=channel, language=langua
ge)
  File "C:\Python27\lib\site-packages\kalite\topic_tools\content_models.py", lin
e 129, in wrapper
    output = map(unparse_model_data, output.dicts())
TypeError: argument 2 to map() must support iteration
```

Somehow I think this may be a missing step when cooking up the sdist...? @benjaoming @rtibbles any insight here?

**Comment by MCGallaspy** [comment]
@benjaoming looking at how the sdist is built. Two things aren't immediately apparent to me:

1. We can specify *which* files to include using the MANIFEST.in template, but we can't specify where files will be installed -- that's up to pip, right? Are there additional ways at install time to specify where files should be installed/permissions for those files?
2. In Windows we used to include an empty, pre-migrated db -- with my PR (https://github.com/learningequality/installers/pull/218) that would change. Do we currently include an empty db (if it exists) in the MANIFEST template? Can we also amend it to include other pre-generated dbs?

To be clear, what I'd like to do is this:

1. Build assets using a make directive (`make assets`), like an empty "main" db using `kalite manage syncdb --noinput` & `kalite manage migrate`, as well as the content item dbs using `kalite manage init_content_items`
2. Include those pre-generated files in the source distribution, and thus install them to a source location at install time.
3. Do the following things at install time or immediately after:
  1. The "main" db must be user-writable -- **if** no database is detected in the user's kalite home directory (`~\.kalite`) **then** at server start do the following:
     1. **If** an empty, pre-generated DB exists in the source, then create a user-writable copy of it in the user's kalite home directory (new functionality)
     2. **If** such a db doesn't exist in the source files, then proceed through setup normally by creating and migrating a DB (current functionality)
  2. The *content* dbs only need to be read-able (right @rtibbles and @aronasorman?). So we simply need to copy them to the source directories and make sure the application is properly configured to use them. (Currently they are not included in the source or improperly configured I think, hence the above error).

**Comment by rtibbles** [comment]
Unless I misunderstood the outcome of the plan that @aronasorman and I had for content packs, then ```init_content_items``` is an interim dev management command to allow development to continue, but will not be used in production. All content databases (including for English) will be prebuilt on the central server.

In addition, the content dbs need to be writable at runtime. This because one significant advantage of the content dbs is the ability to persist the annotation data and *not* have to recreate it at server start - in addition, updates to annotation are done intelligently, rather than wiping everything out and starting again.

**Comment by MCGallaspy** [comment]
Okay, so then what is the plan for production? Will the end-user need to generate a content db at least once before starting the server? Does it make sense to have a "template" db that we can distribute, kind of like how we can/will generate an empty `data.sqlite`?

**Comment by rtibbles** [comment]
No, my latest changes in PR gracefully catches errors from a missing database. So, you'll just have no content until the content database is installed.

**Comment by MCGallaspy** [comment]
Okay, my original motivation for opening this issue was to handle installing assessment items with the Windows installer. The sdist doesn't touch assessment items, as far as I can tell (true @benjaoming?). Just to make sure I now understand this, *after* installation in order to get assessment items working, I need to first get the assessment item zip, unpack it, and then generate the content dbs?

**Comment by rtibbles** [comment]
Yes. Also worth noting that the assessment items db is now folded into the content db as another table.

**Comment by rtibbles** [comment]
In the aforementioned PR #4630

**Comment by benjaoming** [comment]
Urgh, lots of questions, sorry if I can't answer them here. Could we maybe talk about it Monday after standup? It's a little difficult to put in writing.

@MCGallaspy 
MANIFEST.in specifies files in the sourcetree that shall get included in the sourceball (tar.gz or zip). But it's `setup.py` that does the copying of those files to their location.

If we were to include prebuilt databases, the yes the correct order would be:

1. Build them in `make assets`
1. List them in MANIFEST.in.dist for inclusion. In case they aren't built and don't exists, `setup.py sdist` should give a warning
1. If files in MANIFEST.in.dist are *package data*, they will be included automatically. If it is placed outside a package (as with python-packages, but highly discouraged), they need to have a target location in `setup(..., data_files=[...])`.

MANIFEST.in.dist is a convention that I made because the actual MANIFEST.in is generated on-the-fly by custom logic in setup.py!

Btw, at some point, we should have crowdin and `compilemessages` as part of the Makefile, @aronasorman what are your thoughts and would you consider creating an issue for this?

**Comment by MCGallaspy** [comment]
My understanding is that we plan to release content dbs as a downloadable resource, which we can then bundle with our installers. Closing this issue.
